flatpak-import.py

```diff
@@ -0,0 +1,212 @@
+
+    if 'runtime-repo' in metadata_variant.keys():
+        runtime_repo = metadata_variant['runtime-repo']
+    else:
+        runtime_repo = ''
+
+    if 'metadata' in metadata_variant.keys():
+        app_metadata = metadata_variant['metadata']
+    else:
+        app_metadata = ''
+
+    if 'gpg-keys' in metadata_variant.keys():
+        gpg_keys = metadata_variant['gpg-keys']
+    else:
+        gpg_keys = ''
+
+    logging.debug("commit: " + commit)
+    logging.debug("flatpak: " + commit)
+    logging.debug("ref: " + ref)
+    logging.debug("origin: " + origin)
+    logging.debug("runtime_repo: " + runtime_repo)
+    logging.debug("app_metadata:\n---\n" + app_metadata + "---")
+    logging.debug("gpg_keys: " + gpg_keys)
+
+    logging.debug("metadata keys: " + str(metadata_variant.keys()))
+
+
+    # open repository
+
+    repo_file = Gio.File.new_for_path(args.repo)
+    repo = OSTree.Repo(path=repo_file)
+    if os.path.exists(os.path.join(args.repo, 'config')):
+        logging.info('Opening repo at ' + args.repo)
+        repo.open()
+    else:
+        logging.info('Creating archive-z2 repo at ' + args.repo)
+        os.makedirs(args.repo)
+        repo.create(OSTree.RepoMode.ARCHIVE_Z2)
+
+
+    # prepare transaction
+
+    repo.prepare_transaction(None)
+    repo.transaction_set_ref(None, ref, commit)
+
+
+    # execute the delta
+
+    flatpak_file = Gio.File.new_for_path(args.flatpak)
+    repo.static_delta_execute_offline(flatpak_file, False, None)
+
+
+    # verify gpg signature
+
+    if args.gpg_homedir:
+        gpg_homedir = Gio.File.new_for_path(args.gpg_homedir)
+    else:
+        gpg_homedir = None
+    if args.keyring:
+        trusted_keyring = Gio.File.new_for_path(args.keyring)
+    else:
+        trusted_keyring = None
+    gpg_verify_result = repo.verify_commit_ext(commit,
+                                               keyringdir=gpg_homedir,
+                                               extra_keyring=trusted_keyring,
+                                               cancellable=None)
+    # TODO: handle signature requirements
+    if gpg_verify_result and gpg_verify_result.count_valid() > 0:
+        logging.info("valid signature found")
+    else:
+        logging.error("no valid signature")
+        exit(1)
+
+
+    # commit the transaction
+
+    repo.commit_transaction(None)
+
+
+    # grab commit root
+
+    (ret, commit_root, commit_checksum) = repo.read_commit(ref, None)
+    if not ret:
+        logging.critical("commit failed")
+        exit(1)
+    else:
+        logging.debug("commit_checksum: " + commit_checksum)
+
+
+    # compare installed and header metadata, remove commit if mismatch
+
+    metadata_file = Gio.File.resolve_relative_path (commit_root, "metadata");
+    # TODO: GLib-GIO-CRITICAL **: g_file_input_stream_query_info: assertion
+    # 'G_IS_FILE_INPUT_STREAM (stream)' failed
+    (ret, metadata_contents, _) = metadata_file.load_contents(cancellable=None)
+    if ret:
```

Comment:
Here again I expect `load_contents` will raise an exception if it fails. Which is likely to happen in some cases since the `metadata` in the commit root is specific to flatpak and we'd eventually want to upload static deltas from normal ostrees.